File: core/src/main/java/network/aika/queue/steps/Fired.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package network.aika.queue.steps;

import network.aika.activations.Activation;
import network.aika.queue.Phase;
import network.aika.queue.Queue;
import network.aika.queue.Timestamp;
import network.aika.queue.Step;
import network.aika.queue.keys.FieldQueueKey;
import network.aika.utils.ApproximateComparisonValueUtil;
import network.aika.utils.StringUtils;

import static network.aika.queue.Phase.FIRED;

/**
 *
 * @author Lukas Molzberger
 */
public class Fired extends Step<Activation> {

    private final Activation act;

    private double net;

    private int sortValue;

    public Fired(Activation act) {
        this.act = act;
    }

    @Override
    public void createQueueKey(Timestamp timestamp, int round) {
        queueKey = new FieldQueueKey(
                round,
                getPhase(),
                sortValue,
                timestamp
        );
    }

    @Override
    public void process() {
        Activation act = getElement();

        act.setFired();

        // Only once the activation is fired, will it be visible to other neurons.
        act.getBindingSignals()
                .values().
                forEach(bs ->
                        bs.addActivation(act)
                );

        act.linkOutgoing();
    }

    public void updateNet(double net) {
        this.net = net;
        sortValue = ApproximateComparisonValueUtil.convert(net);
    }

    @Override
    public Phase getPhase() {
        return FIRED;
    }

    @Override
    public Activation getElement() {
        return act;
    }

    @Override
    public Queue getQueue() {
        return act.getDocument();
    }

    @Override
    public String toString() {
        return "" + getElement() + " net:" + StringUtils.doubleToString(net);
    }
}

File: src/network/fired.cpp
#include "network/fired.h"
#include "network/activation.h"
#include "fields/queue_key.h"
#include "network/fired_queue_key.h"
#include "network/binding_signal.h"
#include "network/context.h"
#include "network/linker.h"
#include <string>
#include <sstream>
#include <iomanip>

Fired::Fired(Activation* act) : Step(), act(act), net(0.0), sortValue(0) {}

void Fired::createQueueKey(long timestamp, int round) {
    queueKey = new FiredQueueKey(round, getPhase(), getElement(), timestamp);
}

void Fired::process() {
    Activation* act = getElement();
    
    act->setFired();

    // Only once the activation is fired, will it be visible to other neurons.
    for (BindingSignal* bs : act->getBindingSignals()) {
        bs->addActivation(act);
    }

    Linker::linkOutgoing(act);
}

void Fired::updateNet(double net) {
    this->net = net;
    // Simple conversion to int for sort value (handles approximate comparison)
    // The original Java uses ApproximateComparisonValueUtil
    sortValue = static_cast<int>(net * 1000.0); // Scale to preserve 3 decimals of precision
}

const Phase& Fired::getPhase() const {
    return Phase::FIRED;
}

Activation* Fired::getElement() const {
    return act;
}

Queue* Fired::getQueue() const {
    Context* ctx = act->getContext();
    return ctx->getQueue();
}

bool Fired::isQueued() const {
    return getIsQueued();
}

std::string Fired::toString() const {
    // Format the double with fixed precision (3 decimal places)
    std::ostringstream ss;
    ss << std::fixed << std::setprecision(3) << net;
    return getElement()->toString() + " net:" + ss.str();
} 